cpp/bulk_mailer.cc

```diff
@@ -0,0 +1,107 @@
+/** \file    bulk_mailer.cc
+ *  \brief   Sends emails to a list of users.
+ *  \author  Dr. Johannes Ruscheinski
+ */
+
+/*
+    Copyright (C) 2020 Library of the University of Tübingen
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as
+    published by the Free Software Foundation, either version 3 of the
+    License, or (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <vector>
+#include <cstdlib>
+#include "DbConnection.h"
+#include "DbResultSet.h"
+#include "DbRow.h"
+#include "EmailSender.h"
+#include "FileUtil.h"
+#include "VuFind.h"
+#include "util.h"
+
+
+namespace {
+
+
+[[noreturn]] void Usage() {
+    ::Usage("mail_contents sql_statement auxillary_email_address_list reply_to_email_address\n"
+            "\tmail_contents: the file containing the email message\n"
+            "\t\t(the first line will be used as the mail's subject)\n"
+            "\tsql_statement: a query to select the relevant email addresses from the vufind user table\n"
+            "\tauxillary_email_address_list: a path to a plain-text file containing one email address per line\n"
+            "\treply_to_email_address: the email address that will be set as the reply_to field\n");
+}
+
+
+void CollectRecipientsFromSqlTable(const std::string &select_statment, std::vector<std::string> * const recipients) {
+    const auto old_size(recipients->size());
+
+    const auto db_connection(VuFind::GetDbConnection());
+    db_connection->queryOrDie(select_statment);
+    DbResultSet result_set(db_connection->getLastResultSet());
+    DbRow row;
+    while (row = result_set.getNextRow())
+        recipients->emplace_back(row[0]);
+
+    LOG_INFO("Added " + std::to_string(recipients->size() - old_size) + " recipients due to the database query.");
```

Comment:
"Added recipients" could be mistaken for some logic using an INSERT statement. If the function is named "Collect...", why not use this term for logging as well?